File: sally/leetcode/weekly-contest-265/2059. Minimum Operations to Convert Number.cpp
class Solution {
public:
    int min_val = 1e9;
    int visit[1001] = {0,};
    vector<int> num;
    
    void dp(int cur, int goal, int dep){
        if (dep >= min_val) return;
        for(auto target : num){
            int target1 = cur - target;
            if(target1 == goal) {
                min_val = min(min_val, dep);
                return;
            }
            
            int target2 = cur + target;
            if(target2 == goal) {
                min_val = min(min_val, dep);
                return;
            }
            
            int target3 = cur ^ target;
            if(target3 == goal) {
                min_val = min(min_val, dep);
                return;
            }
            
            // valid range
            if(0 <= target1 && target1 <= 1000 && visit[target1] > dep+1){
                visit[target1] = dep+1;
                dp(target1, goal, dep+1);
            }
            if(0 <= target2 && target2 <= 1000 && visit[target2] > dep + 1){
                visit[target2] = dep+1;
                dp(target2, goal, dep+1);
            }
            if(0 <= target3 && target3 <= 1000 && visit[target3] > dep + 1){
                visit[target3] = dep+1;
                dp(target3, goal, dep+1);
            }
        }
    }
    
    int minimumOperations(vector<int>& nums, int start, int goal) {
        // init
        for(auto i :nums) num.push_back(i);
        for(int i = 0; i < 1001; i++) visit[i] = 1001;
        
        // dp
        visit[start] = 0;
        dp(start, goal, 1);
        if(min_val == 1e9) return -1;
        else return min_val;
        
    }
};

File: peacecheejecake/leetcode/2059-minimum-operations-to-convert-number.py
from collections import deque

class Solution:
    def minimumOperations(self, nums: List[int], start: int, goal: int) -> int:
        memo = {start}
        queue = deque([(start, 0)])
        while queue:
            x, deg = queue.popleft()
            if x < 0 or x > 1000:
                continue
            new_nums = set()
            for num in nums:
                cands = x + num, x - num, x ^ num
                for n in cands:
                    if n == goal:
                        return deg + 1
                    
                    if n not in memo:
                        new_nums.add((n, deg + 1))
                        memo.add(n)
                        
            queue.extend(new_nums)
        return -1
            